spec/services/datetime_parser_spec.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+RSpec.describe DatetimeParser do
+  describe "#parse" do
+    context "when the params are valid" do
+      it "returns the parsed date/time" do
+        date_params = { day: "10", month: "01", year: "2019" }
+        expected_time = Time.zone.local(2019, 1, 10, 0, 0)
+
+        datetime = DatetimeParser.new(date: date_params, time: "11:00am").parse
+        expect(datetime).to eq expected_time.change(hour: 11)
+
+        datetime = DatetimeParser.new(date: date_params, time: "9:34").parse
+        expect(datetime).to eq expected_time.change(hour: 9, min: 34)
+
+        datetime = DatetimeParser.new(date: date_params, time: "12:00").parse
+        expect(datetime).to eq expected_time.change(hour: 12)
+
+        datetime = DatetimeParser.new(date: date_params, time: "12:00am").parse
+        expect(datetime).to eq expected_time.change(hour: 0)
+
+        datetime = DatetimeParser.new(date: date_params, time: "6:00 pm").parse
+        expect(datetime).to eq expected_time.change(hour: 18)
+
+        datetime = DatetimeParser.new(date: date_params, time: "23:32").parse
+        expect(datetime).to eq expected_time.change(hour: 23, min: 32)
+
+        datetime = DatetimeParser.new(date: date_params, time: "12:30pm").parse
+        expect(datetime).to eq expected_time.change(hour: 12, min: 30)
+      end
+    end
+
+    context "when the date/time is blank" do
+      it "throws an exception with the issue" do
+        expect { DatetimeParser.new({}).parse }.to raise_error do |error|
+          date_form_message = error.issues.items_for(:schedule_date).first[:text]
+          expect(date_form_message).to eq(I18n.t!("requirements.schedule_date.invalid.form_message"))
+
+          time_form_message = error.issues.items_for(:schedule_time).first[:text]
+          expect(time_form_message).to eq(I18n.t!("requirements.schedule_time.invalid.form_message"))
+        end
+      end
+    end
+
+    context "when the date is invalid" do
+      it "throws an exception with the issue" do
+        params = { date: { day: "10", month: "60", year: "11" } }
+
+        expect { DatetimeParser.new(params).parse }.to raise_error do |error|
+          form_message = error.issues.items_for(:schedule_date).first[:text]
+          expect(form_message).to eq(I18n.t!("requirements.schedule_date.invalid.form_message"))
+        end
+      end
+    end
+
+    context "when the time is invalid" do
+      it "throws an exception with the issue" do
+        params = { date: { day: "10", month: "1", year: "2019" }, time: "13421" }
+
+        expect { DatetimeParser.new(params).parse }.to raise_error do |error|
+          form_message = error.issues.items_for(:schedule_time).first[:text]
+          expect(form_message).to eq(I18n.t!("requirements.schedule_time.invalid.form_message"))
+        end
```

Comment:
It'd be good to have an example here that reflects a value that would be bad if Time.parse was used. 

It's easy to think "why didn't they just use `Time.parse`" without consideration of just how broad a range of values could be accepted and cause odd outcomes.

Reply:
Oof #notmyspec ;-). But yes, this did cross my mind when I first read the implementation - I assumed the '1234' example did that already. I've gone with a different one.